falcon/util/misc.py

```diff
@@ -373,7 +376,16 @@ def secure_filename(filename: str) -> str:
     filename = unicodedata.normalize('NFKD', filename)
     if filename.startswith('.'):
         filename = filename.replace('.', '_', 1)
-    return _UNSAFE_CHARS.sub('_', filename)
+    filename = _UNSAFE_CHARS.sub('_', filename)
+
+    if max_length is not None and len(filename) > max_length:
+        name, ext = os.path.splitext(filename)
+        if max_length < len(ext):
```

Comment:
I don't think it is very important how we handle these edge cases, as long as the behavior is not too complicated, and it is clearly documented what to expect.

That said, maybe we could go with the following logic?
* If the filename fits into the requested _max_length_, obviously return it as-is.
* Split the filename into _stem_ & _extension_, just as you proposed.
* If it is possible to get the filename under the limit by shortening the _stem_ so that at remains non-empty (i.e., at least one character), do it.
* Otherwise, disregard the distinction between the stem and extension, and simply return `filename[:max_length]`.